app/Services/InboundMail/InboundMailEngine.php

```diff
@@ -0,0 +1,268 @@
+        $documents = [];
+        array_push($documents, ...$email->documents);
+        if ($email->body_document !== null)
+            array_push($documents, $email->body_document);
+
+        $expense->saveQuietly();
+
+        $this->saveDocuments($documents, $expense);
+
+        event(new ExpenseWasCreated($expense, $expense->company, Ninja::eventVars(null))); // @turbo124 please check, I copied from API-Controller
+        event('eloquent.created: App\Models\Expense', $expense); // @turbo124 please check, I copied from API-Controller
+    }
+
+    // HELPERS
+    private function processHtmlBodyToDocument(InboundMail $email)
+    {
+
+        if ($email->body !== null)
+            $email->body_document = TempFile::UploadedFileFromRaw($email->body, "E-Mail.html", "text/html");
+
+    }
+    private function validateExpenseSender(Company $company, InboundMail $email)
+    {
+        $parts = explode('@', $email->from);
+        $domain = array_pop($parts);
+
+        // whitelists
+        $email_whitelist = explode(",", $company->inbound_mailbox_whitelist_senders);
+        if (in_array($email->from, $email_whitelist))
+            return true;
+        $domain_whitelist = explode(",", $company->inbound_mailbox_whitelist_domains);
+        if (in_array($domain, $domain_whitelist))
+            return true;
+        $email_blacklist = explode(",", $company->inbound_mailbox_blacklist_senders);
+        if (in_array($email->from, $email_blacklist))
+            return false;
+        $domain_blacklist = explode(",", $company->inbound_mailbox_blacklist_domains);
+        if (in_array($domain, $domain_blacklist))
+            return false;
+
+        // allow unknown
+        if ($company->inbound_mailbox_allow_unknown)
+            return true;
+
+        // own users
+        if ($company->inbound_mailbox_allow_company_users && $company->users()->where("email", $email->from)->exists())
+            return true;
+
+        // from vendors
+        if ($company->inbound_mailbox_allow_vendors && VendorContact::where("company_id", $company->id)->where("email", $email->from)->exists())
+            return true;
+
+        // from clients
+        if ($company->inbound_mailbox_allow_clients && ClientContact::where("company_id", $company->id)->where("email", $email->from)->exists())
+            return true;
+
+        // denie
+        return false;
+    }
+    private function getClient(Company $company, InboundMail $email)
+    {
+        $clientContact = ClientContact::where("company_id", $company->id)->where("email", $email->from)->first();
+        if (!$clientContact)
+            return null;
+
+        return $clientContact->client();
+    }
+    private function getVendor(Company $company, InboundMail $email)
+    {
+        $vendorContact = VendorContact::where("company_id", $company->id)->where("email", $email->from)->first();
+        if (!$vendorContact)
+            return null;
+
+        return $vendorContact->vendor();
+    }
```

Comment:
@turbo124 this implementations could lead to false matching, when multiple contacts on different vendors / clients exists with the same email?

Should we add a logic for handling such cases or ignore it and match with the first found?

Reply:
@paulwer this is true, but I don't think there is a simple solution for this one. The only option I can think of is that we build a route which will allow us to "transfer" a doc to a different client/vendor.

Also, much like bank transactions, i believe we'll want some kind of middleware in the UI to allow the user to adjust the document / reject etc prior to it being assigned anywhere.

Reply:
Currently we save the docs on the expense itself, so the vendor can just be changed on the expense.
The question was more like, how do we notify the user in such case.

Reply:
I will resolve this conversation, as I have desclared, why and how its currently implemented.
Both can be changed from the expense pov.